src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperAttributeList.cs

```diff
@@ -51,112 +50,76 @@ public TagHelperAttributeList(IEnumerable<TagHelperAttribute> attributes)
-                }
-
-                Attributes[index] = value;
+                Items[index] = value;
             }
         }
 
         /// <summary>
-        /// Gets the first <see cref="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/> matching
-        /// <paramref name="name"/>. When setting, replaces the first matching
-        /// <see cref="TagHelperAttribute"/> with the specified <paramref name="value"/> and removes any additional
-        /// matching <see cref="TagHelperAttribute"/>s. If a matching <see cref="TagHelperAttribute"/> is not found,
-        /// adds the specified <paramref name="value"/> to the end of the collection.
-        /// </summary>
+        /// Replaces the first <see cref="TagHelperAttribute"/> whose <see cref="TagHelperAttribute.Name"/> matches the
+        /// specified <paramref name="name"/> and removes any additional matching <see cref="TagHelperAttribute"/>s. If
+        /// a matching <see cref="TagHelperAttribute"/> is not found, adds a <see cref="TagHelperAttribute"/> with
+        /// <paramref name="name"/> and <paramref name="value"/> to the end of the collection.</summary>
         /// <param name="name">
-        /// The <see cref="TagHelperAttribute.Name"/> of the <see cref="TagHelperAttribute"/> to get or set.
+        /// The <see cref="TagHelperAttribute.Name"/> of the <see cref="TagHelperAttribute"/> to set.
         /// </param>
-        /// <returns>The first <see cref="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/> matching
-        /// <paramref name="name"/>.
-        /// </returns>
-        /// <remarks><paramref name="name"/> is compared case-insensitively. When setting,
-        /// <see cref="TagHelperAttribute"/>s <see cref="TagHelperAttribute.Name"/> must be <c>null</c> or
-        /// case-insensitively match the specified <paramref name="name"/>.</remarks>
-        /// <example>
-        /// <code>
-        /// var attributes = new TagHelperAttributeList();
-        ///
-        /// // Will "value" be converted to a TagHelperAttribute with a null Name
-        /// attributes["name"] = "value";
-        ///
-        /// // TagHelperAttribute.Name must match the specified name.
-        /// attributes["name"] = new TagHelperAttribute("name", "value");
-        /// </code>
-        /// </example>
-        public new TagHelperAttribute this[string name]
+        /// <param name="value">
+        /// The <see cref="TagHelperAttribute.Value"/> to set.
+        /// </param>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public void SetAttribute(string name, object value)
         {
-            get
+            if (name == null)
             {
-                if (name == null)
-                {
-                    throw new ArgumentNullException(nameof(name));
-                }
-
-                return base[name];
+                throw new ArgumentNullException(nameof(name));
             }
```

Comment:
No need for a check that's done in the `TagHelperAttribute` constructor. We're fine here because the parameter names match.